docs/advanced_topics/boundblocks_and_values.rst

```diff
@@ -0,0 +1,113 @@
+
+.. code-block:: python
+
+    class HeadingBlock(blocks.CharBlock):
+        class Meta:
+            template = 'blocks/heading.html'
+
+where ``blocks/heading.html`` consists of:
+
+.. code-block:: html+django
+
+    <h1>{{ value }}</h1>
+
+This gives us a block that behaves as an ordinary text field, but wraps its output in ``<h1>`` tags whenever it is rendered:
+
+.. code-block:: python
+
+    class BlogPage(Page):
+        body = StreamField([
+            # ...
+            ('heading', HeadingBlock()),
+            # ...
+        ])
+
+.. code-block:: html+django
+
+    {% load wagtailcore_tags %}
+
+    {% for block in page.body %}
+        {% if block.block_type == 'heading' %}
+            {% include_block block %}  {# This block will output its own <h1>...</h1> tags. #}
+        {% endif %}
+    {% endfor %}
+
+This kind of arrangement - a value that supposedly represents a plain text string, but has its own custom HTML representation when output on a template - would normally be a very messy thing to achieve in Python, but it works here because the items you get when iterating over a StreamField are not actually the 'native' values of the blocks. Instead, each item is returned as an instance of ``BoundBlock`` - an object that represents the pairing of a value and its block definition. By keeping track of the block definition, a ``BoundBlock`` always knows which template to render. To get to the underlying value - in this case, the text content of the heading - you would need to access ``block.value``. Indeed, if you were to output ``{% include_block block.value %}`` on the page, you would find that it renders as plain text, without the ``<h1>`` tags.
+
+(More precisely, the items returned when iterating over a StreamField are instances of a class ``StreamChild``, which provides the ``block_type`` property as well as ``value``.)
+
+Experienced Django developers may find it helpful to compare this to the ``BoundField`` class in Django's forms framework, which represents the pairing of a form field value with its corresponding form field definition, and therefore knows how to render the value as an HTML form field.
+
+Most of the time, you won't need to worry about these internal details; Wagtail will use the template rendering wherever you would expect it to. However, there are certain cases where the illusion isn't quite complete - namely, when accessing children of a ``ListBlock`` or ``StructBlock``. In these cases, there is no ``BoundBlock`` wrapper, and so the item cannot be relied upon to know its own template rendering. For example, consider the following setup, where our ``HeadingBlock`` is a child of a StructBlock:
```

Comment:
Missing backticks around ``StructBlock``

Reply:
As per https://github.com/wagtail/wagtail/pull/6771#issuecomment-778244981, I think we should standardise on _not_ using backticks when we're referring to StructBlocks as a concept, e.g. "a child of a StructBlock". But since we're just moving existing text around here, that's a discussion we should have elsewhere...